collector/hot_threads.go

```diff
@@ -0,0 +1,246 @@
+	HotThreadsFailureMetrics HotThreadsStepFailureMetric
+
+	jsonParseFailures prometheus.Counter
+}
+
+type HotThreadsMetric struct {
+	Type   prometheus.ValueType
+	Desc   *prometheus.Desc
+	Value  func(HotThreadsExp float64) float64
+	Labels func(HotThreadsDataNode string, HotThreadsName, HotThreadsId string) []string
+}
+
+type HotThreadsStepFailureMetric struct {
+	Type   prometheus.ValueType
+	Desc   *prometheus.Desc
+	Value  func(HotThreadsExp int64) float64
+	Labels func(HotThreadsIndex string, HotThreadsPolicy string, action string, step string) []string
+}
+
+func getEnv(key, defaultVal string) string {
+	value, exists := os.LookupEnv(key)
+	if !exists {
+		value = defaultVal
+	}
+	return value
+}
+
+// NewHotThreadsExplain defines HotThreads Prometheus metrics
+func NewHotThreads(logger log.Logger, url *url.URL) *HotThreads {
+	return &HotThreads{
+		logger: logger,
+		url:    url,
+
+		HotThreadsMetrics: HotThreadsMetric{
+			Type: prometheus.GaugeValue,
+			Desc: prometheus.NewDesc(
+				prometheus.BuildFQName(namespace, "hot_threads", "cpu_usage_percentage"),
+				"Hot Threads cpu usage on data nodes",
+				defaultHotThreadsLabels, nil,
+			),
+			Value: func(HotThreadsCpuPercentage float64) float64 {
+				return float64(HotThreadsCpuPercentage)
+			},
+			Labels: func(HotThreadsDataNode string, HotThreadsName, HotThreadsId string) []string {
+				return []string{HotThreadsDataNode, HotThreadsName, HotThreadsId}
+			},
+		},
+
+		jsonParseFailures: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: prometheus.BuildFQName(namespace, "hot_threads", "json_parse_failures"),
+			Help: "Number of errors while parsing JSON.",
+		}),
+	}
+}
+
+// Describe HotThreads
+func (s *HotThreads) Describe(ch chan<- *prometheus.Desc) {
+	ch <- s.jsonParseFailures.Desc()
+	ch <- s.HotThreadsMetrics.Desc
+}
+
+func (s *HotThreads) getAndParseURL(u *url.URL, hotThreads *[]HotThreadsRsp) error {
+	res, err := http.Get(u.String())
+	if err != nil {
+		return fmt.Errorf("failed to get from %s://%s:%s%s: %s",
+			u.Scheme, u.Hostname(), u.Port(), u.Path, err)
+	}
+
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		_ = level.Warn(s.logger).Log(
+			"msg", "failed to get resp body",
+			"err", err,
+		)
+	}
+
+	defer func() {
+		err = res.Body.Close()
+		if err != nil {
+			_ = level.Warn(s.logger).Log(
+				"msg", "failed to close http.Client",
+				"err", err,
+			)
+		}
+	}()
+
+	if res.StatusCode != http.StatusOK {
+		return fmt.Errorf("HTTP Request failed with code %d", res.StatusCode)
+	}
+
+	sb := string(body)
+	hotThreadsNodeOp := strings.Split(string(sb), NODE_OUTPUT_SEPERATOR)
+
+	for _, nodeData := range hotThreadsNodeOp {
+		nodeName := strings.Trim(strings.Split(nodeData, "}")[0], " {")
+
+		hotThreadsOpRegex := regexp.MustCompile(HOT_THREADS_OP_REGEX)
+		allHotThreads := hotThreadsOpRegex.FindAllString(nodeData, -1)
+		cpuPercentageRegex := regexp.MustCompile(CPU_PERCENTAGE_REGEX)
+
+		for _, v := range allHotThreads {
```

Comment:
The logic here seems like some complex parsing. I think it would be best for this to be in a separate function with good unit tests. The signature of the function and name should provide a clear definition of what the functionality is. 

Is there not an alternative to this parsing? What format is the data in?